### PR TITLE
Porting wolfssl into ESP-IDF development framework

### PR DESCRIPTION
**Pull request by miyazakh**
Port wolfssl into ESP-IDF development framework, including benchmark, cyrpt-test and simple sever/client for estimation

### CODE REVIEW

File: IDE/Espressif/ESP-IDF/examples/wolfssl_benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+The Example contains of wolfSSL benchmark program.
+
+1. "make menuconfig" to configure the program.
+ 1-1. Example Configuration ->
+     BENCH_ARG : argument that you want to use. Default is "-lng 0"
+                 The list of argument can be find in help.
+
+When you want to run the benchmark program
+1. "make flash" to compile and load the firemware
```
Comment (cconlon) [inline IDE/Espressif/ESP-IDF/examples/wolfssl_benchmark/README.md]:
```suggestion
1. "make flash" to compile and load the firmware
```

File: IDE/Espressif/ESP-IDF/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+## Requirements
+ 1. ESP-IDF development framework
+    [https://docs.espressif.com/projects/esp-idf/en/latest/get-started/]
+    Note: This expects to use Linux version.
+     
+## Setup
+ 1. Run *setup.sh* to deploy files into ESP-IDF tree
+ 2. Find Wolfssl files at /path/to/esp-idf/components/wolfssl/
+ 3. Find Example programs under /path/to/esp-idf/examples/protocols/wolfssl_xxx
+ 4. Uncomment out #define WOLFSSL_ESPIDF in /path/to/wolfssl/wolfssl/wolfcrypt/setting.h
```
Comment (cconlon) [inline IDE/Espressif/ESP-IDF/README.md]:
```suggestion
 4. Uncomment out #define WOLFSSL_ESPIDF in /path/to/wolfssl/wolfssl/wolfcrypt/settings.h
```

File: IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#wolfSSL Example
+
+The Example contains of wolfSSL simple server.
```
Comment (cconlon) [inline IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md]:
```suggestion
This example contains a wolfSSL simple server.
```

File: IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+1. "make menuconfigure" to configure the project
+  1-1. Example Configuration ->
+       WIFI SSID : your own FIFI, which is connected to the Internet.(default is "myssid")
```
Comment (cconlon) [inline IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md]:
```suggestion
       WIFI SSID : your own WIFI, which is connected to the Internet.(default is "myssid")
```

File: wolfcrypt/benchmark/benchmark.c [hunk 5]
```diff
@@ -4986,10 +4986,14 @@ void bench_ed25519KeySign(void)
     /* declared above at line 239 */
     /* extern   double current_time(int reset); */
 
-#elif defined FREERTOS
+#elif defined(FREERTOS) || defined(WOLFSSL_ESPIDF)
```
Comment (cconlon) [inline wolfcrypt/benchmark/benchmark.c]: I think "task.h" is a FreeRTOS header, and not specific to the ESP-IDF.  It looks like your settings.h automatically defines `FREERTOS` when `WOLFSSL_ESPIDF` is defined.  This probably means we can remove the check here for `WOLFSSL_ESPIDF` right?

File: wolfcrypt/src/random.c [hunk 6]
```diff
@@ -2060,6 +2060,24 @@ int wc_GenerateSeed(OS_Seed* os, byte* output, word32 sz)
...
+            int i;
+            
+            for (i = 0; i< sz; i++) {
+               output[i] =  esp_random( );
+            }
+        
+            return 0;
+        }
+    #else
+        #define USE_TEST_GENSEED
```
Comment (cconlon) [inline wolfcrypt/src/random.c]: Let's remove this.  We want the user's compiler to give them a warning if they need to write their own GenerateSeed() which we do not support yet.  Having this here could make it easier for people to ship without a valid GenerateSeed() implemented.

File: wolfssl/wolfcrypt/settings.h [hunk 7]
```diff
@@ -175,6 +175,9 @@
 /* Uncomment next line if building for using Apache mynewt */
 /* #define WOLFSSL_APACHE_MYNEWT */
 
+/* Uncomment next line if building for using ESP-IDF */
+/* #define WOLFSSL_ESPIDF */
+
```
Comment (cconlon) [inline wolfssl/wolfcrypt/settings.h]: We should have another define in here for `WOLFSSL_ESPWROOM32`, right?  If we don't list it here, there's a chance people will overlook it.

File: wolfssl/wolfcrypt/settings.h [hunk 8]
```diff
@@ -216,6 +219,23 @@
...
+    #define SIZEOF_LONG_LONG 8
+    #define NO_WOLFSSL_DIR
+    #define WOLFSSL_NO_CURRDIR
+
+    #define TFM_TIMING_RESISTANT
+    #define ECC_TIMING_RESISTANT
+    #define WC_RSA_BLINDING
+#if !defined(WOLFSSL_USER_SETTINGS)
+    #define WOLFSSL_ESPWROOM32
+    #define HAVE_ECC
```
Comment (cconlon) [inline wolfssl/wolfcrypt/settings.h]: I'm not sure we want to force all ESP users to have ECC enabled by default.  Can we remove this and put it in our example user_settings.h instead?

File: wolfssl/wolfcrypt/wc_port.h [hunk 9]
```diff
@@ -365,7 +365,10 @@ WOLFSSL_API int wolfCrypt_Cleanup(void);
     WOLFSSL_API int wc_ReadDirNext(ReadDirCtx* ctx, const char* path, char** name);
     WOLFSSL_API void wc_ReadDirClose(ReadDirCtx* ctx);
 #endif /* !NO_WOLFSSL_DIR */
-
+#else
+    #ifdef WOLFSSL_ESPIDF
+        #include <stdio.h>
```
Comment (cconlon) [inline wolfssl/wolfcrypt/wc_port.h]: Do we still need this include for `stdio.h`, now that you have logging functionality integrated?

File: wolfssl/wolfcrypt/settings.h (hunk 8 above)
Comment (cconlon) [reply wolfssl/wolfcrypt/settings.h]: Disregard, based on conversation.

File: IDE/Espressif/ESP-IDF/README.md (hunk 2 above)
Comment (miyazakh) [reply IDE/Espressif/ESP-IDF/README.md]: Thanks, I fixed it.

File: IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md (hunk 3 above)
Comment (miyazakh) [reply IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md]: Thanks, I fixed it.

File: IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md (hunk 4 above)
Comment (miyazakh) [reply IDE/Espressif/ESP-IDF/examples/wolfssl_server/README.md]: Thanks, I fixed it.

File: wolfcrypt/benchmark/benchmark.c (hunk 5 above)
Comment (miyazakh) [reply wolfcrypt/benchmark/benchmark.c]: Yes, right. I am going to remove the ESP-IDF.

File: wolfcrypt/src/random.c (hunk 6 above)
Comment (miyazakh) [reply wolfcrypt/src/random.c]: OK. I am going to remove this.

File: wolfssl/wolfcrypt/settings.h (hunk 7 above)
Comment (miyazakh) [reply wolfssl/wolfcrypt/settings.h]: I agree with you. I am going to add WOLFSSL_ESPWROOM32 define in the file and remove the define in user_settings.h.

File: wolfssl/wolfcrypt/wc_port.h (hunk 9 above)
Comment (miyazakh) [reply wolfssl/wolfcrypt/wc_port.h]: No, we don't need it anymore. I am going to remove this.

File: wolfcrypt/benchmark/benchmark.c (hunk 5 above)
Comment (cconlon) [reply wolfcrypt/benchmark/benchmark.c]: Thanks

File: wolfcrypt/src/random.c (hunk 6 above)
Comment (cconlon) [reply wolfcrypt/src/random.c]: Thanks

File: wolfssl/wolfcrypt/settings.h (hunk 7 above)
Comment (cconlon) [reply wolfssl/wolfcrypt/settings.h]: Thanks

**Comment by cconlon** [comment]
Thanks, please rebase on master to fix the failing Jenkins issue.

**Comment by cconlon** [comment]
Thanks!

**Pull request merged**